### PR TITLE
update mm-sink tutorial and plot

### PR DESCRIPTION
**Pull request by zoepiran**
Updates MM-OT tutorial visualization and adds MM-OT plotting utils.

Updates in files:
- `docs/references.bib:` add citation `lin:22`
- `docs/tutorials/linear/600_mmsink.ipynb`: update visuals and add comparison to MM (det. solution)
- `src/ott/tools/plot.py`: add class `PlotMM`

### CODE REVIEW

File: src/ott/tools/plot.py [hunk 1]
```diff
@@ -245,3 +260,159 @@ def animate(
...
+    assert self._n < jnp.prod(
+        jnp.array(n_s)
+    ), "Intended number of tuples too large."
+
+    # Extract top_k largest entries in the tensor, and their indices.
+    val, idx = jax.lax.top_k(ot.tensor.ravel(), self._top_k)
+    indices = jnp.unravel_index(idx, n_s)
+
+    # Setttings for plot
+    markers = "svopxdh"
```
Comment (michalk8) [inline src/ott/tools/plot.py:324]: I'd add this option to `__init__`, as this will fail if there are more than 7 marginals (unlikely, but still).

File: src/ott/tools/plot.py [hunk 2]
```diff
@@ -245,3 +260,159 @@ def animate(
...
+        title=title
+    )
+
+    self._patches = []
+    self._points = []
+    self._n, self._k, self._top_k = None, None, None
+
+  def __call__(self, ot: mmsinkhorn.MMSinkhornOutput) -> List["plt.Artist"]:
+    """Plot 2-D couplings. does not support higher dimensional."""
+    n_s = [len(ot.x_s[i]) for i in range(len(ot.x_s))]
```
Comment (michalk8) [inline src/ott/tools/plot.py:314]: Would use `n_s = ot.shape`.

File: src/ott/tools/plot.py [hunk 3]
```diff
@@ -245,3 +260,159 @@ def animate(
...
+      points = [ot.x_s[i][indices[i][j], :] for i in range(self._k)]
+      # reorder to ensure polygons have maximal area
+      points = [points[i] for i in ccworder(jnp.array(points))]
+      alpha = 0.6 if j < self._n else alphas[j - self._n]
+      self._patches[j].set_xy(points)
+      self._patches[j].set_color(self._cmap[j > self._n])
+      self._patches[j].set_alpha(alpha)
+
+    for i in range(self._k):
+      for j, val in enumerate(ot.x_s[i]):
```
Comment (michalk8) [inline src/ott/tools/plot.py:385]: As mentioned above, if possible, I would avoid iterating over the points.

File: src/ott/tools/plot.py [hunk 4]
```diff
@@ -245,3 +260,159 @@ def animate(
...
+
+  def __call__(self, ot: mmsinkhorn.MMSinkhornOutput) -> List["plt.Artist"]:
+    """Plot 2-D couplings. does not support higher dimensional."""
+    n_s = [len(ot.x_s[i]) for i in range(len(ot.x_s))]
+    assert self._n < jnp.prod(
+        jnp.array(n_s)
+    ), "Intended number of tuples too large."
+
+    # Extract top_k largest entries in the tensor, and their indices.
+    val, idx = jax.lax.top_k(ot.tensor.ravel(), self._top_k)
```
Comment (michalk8) [inline src/ott/tools/plot.py:320]: I'd consider passing `top_k` instead of saving it. Then in the notebook, you can use the `__call__` method directly for static plotting as well, instead of re-defining the functions.

File: src/ott/tools/plot.py [hunk 5]
```diff
@@ -245,3 +260,159 @@ def animate(
...
+      transports: Sequence[Transport],
+      titles: Optional[Sequence[str]] = None,
+      frame_rate: float = 10.0,
+      **kwargs
+  ) -> "animation.FuncAnimation":
+    """Make an animation from several transports."""
+    self._k = len(transports[0].tensor.shape)
+    self._n = transports[0].tensor.shape[0]
+
+    self._top_k = kwargs.pop("top_k", self._n)
```
Comment (michalk8) [inline src/ott/tools/plot.py:407]: Would not save this, instead pass it to `__call__/update`.

File: src/ott/tools/plot.py [hunk 6]
```diff
@@ -245,3 +260,159 @@ def animate(
...
+    indices = jnp.unravel_index(idx, n_s)
+
+    alphas = np.linspace(0.6, 0.2, self._top_k - self._n)
+    for j in range(self._top_k):
+      points = [ot.x_s[i][indices[i][j], :] for i in range(self._k)]
+      # reorder to ensure polygons have maximal area
+      points = [points[i] for i in ccworder(jnp.array(points))]
+      alpha = 0.6 if j < self._n else alphas[j - self._n]
+      self._patches[j].set_xy(points)
+      self._patches[j].set_color(self._cmap[j > self._n])
```
Comment (michalk8) [inline src/ott/tools/plot.py:381]: Please add a comment explaining what this does.

File: src/ott/tools/plot.py [hunk 7]
```diff
@@ -245,3 +260,159 @@ def animate(
...
+    indices = jnp.unravel_index(idx, n_s)
+
+    # Setttings for plot
+    markers = "svopxdh"
+
+    for j in range(self._top_k):
+      points = [ot.x_s[i][indices[i][j], :] for i in range(self._k)]
+      points = [points[i] for i in ccworder(jnp.array(points))]
+      alphas = np.linspace(0.6, 0.2, self._top_k - self._n)
+      alpha = 0.6 if j < self._n else alphas[j - self._n]
```
Comment (michalk8) [inline src/ott/tools/plot.py:330]: Maybe instead of the hard-coded `0.6`, let's use `self._alpha` (also in other places)?

File: src/ott/tools/plot.py [hunk 8]
```diff
@@ -245,3 +260,159 @@ def animate(
...
+        ax=ax,
+        cmap=cmap,
+        scale_alpha_by_coupling=scale_alpha_by_coupling,
+        alpha=alpha,
+        title=title
+    )
+
+    self._patches = []
+    self._points = []
+    self._n, self._k, self._top_k = None, None, None
```
Comment (michalk8) [inline src/ott/tools/plot.py:310]: As discussed below, don't think we need to store this.

File: src/ott/tools/plot.py (hunk 5 above)
Comment (michalk8) [reply src/ott/tools/plot.py:407]: Also, why `top_k` should be by default `self._n`?

**Review by michalk8 — CHANGES_REQUESTED** [review]

File: docs/references.bib [hunk 9]
```diff
@@ -1016,3 +1016,13 @@ @misc{kassraie:24
   title       = {Progressive Entropic Optimal Transport Solvers},
   year        = {2024},
 }
+
+@article{lin:22,
```
[the remark this answers is not in the document]
Comment (zoepiran) [reply docs/references.bib:1020]: the JMLR citation doesnt include it, we can manually add:
`url={https://www.jmlr.org/papers/volume23/19-843/19-843.pdf},`
should I?

File: src/ott/tools/plot.py [hunk 10]
```diff
@@ -13,23 +13,37 @@
...
 from ott.geometry import pointcloud
 from ott.solvers.linear import sinkhorn, sinkhorn_lr
 from ott.solvers.quadratic import gromov_wasserstein
 
 try:
+  import matplotlib
+  import matplotlib.patches as ptc
   import matplotlib.pyplot as plt
   from matplotlib import animation
+  matplotlib.rcParams["animation.embed_limit"] = 2 ** 128
```
[the remark this answers is not in the document]
Comment (zoepiran) [reply src/ott/tools/plot.py:31]: to save large obj. but not necessary indeed.

File: src/ott/tools/plot.py [hunk 11]
```diff
@@ -245,3 +260,159 @@ def animate(
...
+
+    return self._points + self._patches
+
+  def update(
+      self,
+      ot: mmsinkhorn.MMSinkhornOutput,
+      title: Optional[str] = None
+  ) -> List["plt.Artist"]:
+    """Update a plot with a transport instance."""
+    n_s = [len(ot.x_s[i]) for i in range(len(ot.x_s))]
```
[the remark this answers is not in the document]
Comment (zoepiran) [reply src/ott/tools/plot.py:367]: these are outputs of `mmsinkhorn.MMSinkhornOutput`, they do not follow same API.

[the remark this answers is not in the document]
Comment (zoepiran) [reply src/ott/tools/plot.py:367]: 🙏🏾

File: src/ott/tools/plot.py [hunk 12]
```diff
@@ -245,3 +260,159 @@ def animate(
...
+
+  def animate(
+      self,
+      transports: Sequence[Transport],
+      titles: Optional[Sequence[str]] = None,
+      frame_rate: float = 10.0,
+      **kwargs
+  ) -> "animation.FuncAnimation":
+    """Make an animation from several transports."""
+    self._k = len(transports[0].tensor.shape)
```
[the remark this answers is not in the document]
Comment (zoepiran) [reply src/ott/tools/plot.py:404]: 🙏🏾

[the remark this answers is not in the document]
Comment (zoepiran) [reply src/ott/tools/plot.py:404]: removed

File: src/ott/tools/plot.py [hunk 13]
```diff
@@ -245,3 +260,159 @@ def animate(
...
+    self._n = transports[0].tensor.shape[0]
```
[the remark this answers is not in the document]
Comment (zoepiran) [reply src/ott/tools/plot.py:405]: removed

File: src/ott/tools/plot.py [hunk 14]
```diff
@@ -245,3 +260,159 @@ def animate(
...
+    )
+
+    self._patches = []
+    self._points = []
+    self._n, self._k, self._top_k = None, None, None
+
+  def __call__(self, ot: mmsinkhorn.MMSinkhornOutput) -> List["plt.Artist"]:
+    """Plot 2-D couplings. does not support higher dimensional."""
+    n_s = [len(ot.x_s[i]) for i in range(len(ot.x_s))]
+    assert self._n < jnp.prod(
```
[the remark this answers is not in the document]
Comment (zoepiran) [reply src/ott/tools/plot.py:315]: removed

File: src/ott/tools/plot.py [hunk 15]
```diff
@@ -245,3 +260,159 @@ def animate(
...
+    return self._points + self._patches
+
+  def update(
+      self,
+      ot: mmsinkhorn.MMSinkhornOutput,
+      title: Optional[str] = None
+  ) -> List["plt.Artist"]:
+    """Update a plot with a transport instance."""
+    n_s = [len(ot.x_s[i]) for i in range(len(ot.x_s))]
+    assert self._n < jnp.prod(jnp.array(n_s)), \
```
[the remark this answers is not in the document]
Comment (zoepiran) [reply src/ott/tools/plot.py:368]: removed

File: src/ott/tools/plot.py [hunk 16]
```diff
@@ -245,3 +260,159 @@ def animate(
...
+
+    for i in range(self._k):
+      for j, val in enumerate(ot.x_s[i]):
+        idx = np.ravel_multi_index((i, j), (self._k, self._n))
+        self._points[idx].set_offsets(val)
+
+    if title is not None:
+      self.ax.set_title(title)
+
+    self.ax.set_ylim(-2.5e-2, 1 + 2.5e-2)
```
[the remark this answers is not in the document]
Comment (zoepiran) [reply src/ott/tools/plot.py:392]: no; `ccworder` is only for sorting. 
because i wanted to fix the limits and we have normalized data. changed a bit.

[the remark this answers is not in the document]
Comment (zoepiran) [reply src/ott/tools/plot.py:392]: i think the change is ugly :(

File: src/ott/tools/plot.py [hunk 17]
```diff
@@ -245,3 +260,159 @@ def animate(
...
+          fill=True,
+          linewidth=2,
+          color=self._cmap[j > self._n],
+          alpha=alpha,
+          zorder=-j,
+      )
+      self._patches.append(self.ax.add_patch(points))
+
+    for i in range(self._k):
+      for j, val in enumerate(ot.x_s[i]):
```
[the remark this answers is not in the document]
Comment (zoepiran) [reply src/ott/tools/plot.py:342]: its the pain in the update indeed--can you figure the enumeration easily?

File: src/ott/tools/plot.py [hunk 18]
```diff
@@ -245,3 +260,159 @@ def animate(
...
+    # Extract top_k largest entries in the tensor, and their indices.
+    val, idx = jax.lax.top_k(ot.tensor.ravel(), self._top_k)
+    indices = jnp.unravel_index(idx, n_s)
+
+    # Setttings for plot
+    markers = "svopxdh"
+
+    for j in range(self._top_k):
+      points = [ot.x_s[i][indices[i][j], :] for i in range(self._k)]
+      points = [points[i] for i in ccworder(jnp.array(points))]
```
[the remark this answers is not in the document]
Comment (zoepiran) [reply src/ott/tools/plot.py:328]: Not sure i follow -> you need to extract from the indices object `top_k` vals

File: src/ott/tools/plot.py (hunk 1 above)
Comment (zoepiran) [reply src/ott/tools/plot.py:324]: i did something ugly; wdyt?

```
marker=markers[i % len(markers)],
c="black" if i < len(markers) else "grey",
```

File: src/ott/tools/plot.py (hunk 2 above)
Comment (zoepiran) [reply src/ott/tools/plot.py:314]: modified

File: src/ott/tools/plot.py (hunk 3 above)
Comment (zoepiran) [reply src/ott/tools/plot.py:385]: not sure how. will think of it

File: src/ott/tools/plot.py (hunk 5 above)
Comment (zoepiran) [reply src/ott/tools/plot.py:407]: because this is the minimal set of mapping, for each point single polygon

File: src/ott/tools/plot.py (hunk 8 above)
Comment (zoepiran) [reply src/ott/tools/plot.py:310]: removed

**Comment by zoepiran** [comment]
<p>genious</p>

---

 <a href='https://app.reviewnb.com/ott-jax/ott/pull/572/discussion/#comment-2296263401'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

**Comment by zoepiran** [comment]
<p>horrible phrasing</p>

---

 <a href='https://app.reviewnb.com/ott-jax/ott/pull/572/discussion/#comment-2296263414'>View entire conversation</a> on ReviewNB<div id="RAICReplyForNotification-ReviewNBCommentContext-DoNotDelete"></div>

File: src/ott/tools/plot.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (michalk8) [reply src/ott/tools/plot.py:343]: Will take a look!

**Review by michalk8 — APPROVED** [review]
LGTM, will wait until the CI passes and then merge! Thanks @zoepiran !
